### DISCUSSION TITLE
Feature Request: Allow Matrix Fields to be Visual / Drag-and-Dropped

### DISCUSSION THREAD

**Discussion by RitterKnightCreative** [craftcms/cms#8540] [Ideas]
One of the things about Matrix that's not as intuitive as it could be is adding and rearranging blocks. You either have to find the block you want and add a block above it or add a block from the buttons on bottom of the screen and then rearrange it in the stack. With a lot of different blocks, it can get unruly quickly.

Matrix is also very text heavy.  One of the things that might help is to allow developers to choose from different "icons" (or even upload, say,  a custom SVG) and allow those icons to represent the different matrix blocks one could choose from. Maybe even allow them to show just the icon when collapsed.

It would be a cold day in hell before we switched back to WordPress, but I've been poking in and out periodically of that ecosystem and there is definitely some interesting ideas floating around regarding content editing. Front end editors like [Thrive Architect](https://thrivethemes.com/architect/), [Elementator](https://elementor.com), and WordPress' own upcoming Gutenberg editor come to mind.  

Those first two third party plugins in particular allow you to drag and drop prebuilt "blocks" into an inline editor.  I'm not convinced an "inline" editor would be a  great for  thing Craft but being able to create a new matrix block by dragging a button up and down in the stack would be intuitive.  Another feature could be "pinning" frequently used blocks to the top and then drag them into the matrix stack.

**Comment by brandonkelly** [comment]
Thanks for the suggestions! We do have some Matrix improvements planned for 3.x, so we’ll keep this in mind.

**Comment by RitterKnightCreative** [comment]
You bet, looking forward to 'em.

**Comment by hiasl** [comment]
I think this topic is still very hot. Though the [Spoon extension](https://github.com/angell-co/Spoon) does a lot of good things, I would really love to see drag'n'drop as an alternative GUI for adding matrix blocks. Combined with grouping as in Spoon would be totally cool.
